src/crab/linear_expression.hpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include <map>
+#include "variable.hpp"
+
+using number_t = crab::z_number;
+using variable_t = crab::variable_t;
+
+// A linear expression is of the form: Ax + By + Cz + ... + N.
+// That is, a sum of terms where each term is either a
+// coefficient * variable, or simply a coefficient
+// (of which there is only one such term).
+class linear_expression_t final {
+
+    // Use a map for the variable terms to simplify adding two expressions
+    // with the same variable.
+    using variable_terms_t = std::map<variable_t, number_t>;
+
+  private:
+    variable_terms_t _variable_terms;
+    number_t _constant_term;
+
+    // Get the coefficient for a given variable, which is 0 if it has no term in the expression.
+    number_t coefficient_of(const variable_t& variable) const {
+        auto it = _variable_terms.find(variable);
+        if (it == _variable_terms.end()) {
+            return 0;
+        }
+        return (*it).second;
+    }
+
+  public:
+    linear_expression_t(signed long long int coefficient) { _constant_term = coefficient; }
+    linear_expression_t(const number_t& coefficient) { _constant_term = coefficient; }
+    linear_expression_t(const variable_t& variable) { _variable_terms[variable] = 1; }
+    linear_expression_t(const number_t& coefficient, const variable_t& variable) { _variable_terms[variable] = coefficient; }
+    linear_expression_t(const variable_terms_t& variable_terms, const number_t& constant_term) {
+        _variable_terms = variable_terms;
+        _constant_term = constant_term;
+    }
+
+    // Allow a caller to access individual terms.
+    const variable_terms_t& variable_terms() const { return _variable_terms; }
+    const number_t& constant_term() const { return _constant_term; }
+
+    // Multiply a linear expression by a constant.
+    [[nodiscard]] linear_expression_t operator*(const number_t& constant) const {
+        variable_terms_t variable_terms;
+        for (const auto& [variable, coefficient] : _variable_terms) {
+            variable_terms[variable] = coefficient * constant;
```

Comment:
I think `.insert()` is better suited here.

Reply:
After reading https://stackoverflow.com/questions/17172080/insert-vs-emplace-vs-operator-in-c-map, I updated it to use emplace()